### PR TITLE
🐛 Source Amplitude: Fix an infinite loop

### LINKED ISSUE #13500: Amplitude stuck in infinite loop generating duplicate events
## Environment
- **Airbyte version**:  0.39.10-alpha
- **OS Version / Instance**: Ubuntu 20.04.4 LTS
- **Deployment**: Docker
- **Source Connector and version**: `airbyte/source-amplitude:0.1.8`
- **Destination Connector and version**: `airbyte/destination-postgres:0.3.20`
- **Step where error happened**: Sync job

## Current Behavior

Amplitude connector is stuck in an infinite loop recording every event over and over again, hence never finishing a sync and ending up filling all available disk space. It's reproducible 100% of the time in "incremental + deduped history" and "full refresh + overwrite" modes.

```
SELECT COUNT(*), _airbyte_data->>'$insert_id' FROM amplitude2._airbyte_tmp_utw_events GROUP BY _airbyte_data->>'$insert_id' HAVING COUNT(*) > 1;

count  $insert_id
38  27432c06-6751-41e8-b3e6-df68698868b2
38  e1967d9c-5432-4144-9b2c-f7fe5a173f15
38  c0fe1ec7-0fbd-4d4f-a8b4-513c4bb22a71
38  46bab8e1-34bd-4521-b72c-428498450c5d
38  6f40b324-5e56-4451-9521-c6a15d5840c9
38  2a6a7553-9427-4119-826e-16681b821629
38  26b52fcf-49d4-4506-88c2-2cf60d7ebf53
...
```

### PR DESCRIPTION
**Pull request by b4stien**
## What

This PR fixes [this issue](https://github.com/airbytehq/airbyte/issues/13500) by implementing a fix to the problem [described here](https://github.com/airbytehq/airbyte/issues/13500#issuecomment-1150729342).

## How

- Disable the use of HTTP pagination for Amplitude events (the stream slices are used to split incoming data into small chunks), this is the change to `Events.next_page_token()`
- Fixing the boundary of the stream slices to not request the first hour of a day twice, this is the change to `Events.stream_slices()`

## 🚨 User Impact 🚨

There is no breaking change, the issue is fixed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<strong>Integration</strong>

Screenshot 2022-06-10 at 09 45 29

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

**Comment by b4stien** [comment]
I've mentionned it in the issue: you currently have this bug in prod (see https://cloud.airbyte.io/workspaces/9ac913ae-9364-47a0-a48d-60ab59022fa2/connections/3fe3bca0-c929-4699-a134-ab9fb2cad8dd/status), and my PR fixed it on my self-hosted instance.

**Comment by harshithmullapudi** [comment]
Hey thanks for the contribution. Could you help us with 
1. Share the screenshot of integration tests
2. Also you can follow the documentation here https://docs.airbyte.com/connector-development/#updating-an-existing-connector to make changes to certain files above before we merge this

**Comment by b4stien** [comment]
@harshithmullapudi done

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/api.py [hunk 1]
```diff
@@ -161,13 +161,15 @@ def stream_slices(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Ite
...
+            slice_end = next_start.subtract(hours=1)
             slices.append(
                 {
                     "start": start.strftime(self.date_template),
-                    "end": self._get_end_date(start).strftime(self.date_template),
+                    "end": slice_end.strftime(self.date_template)
                 }
             )
-            start = start.add(**self.time_interval)
+            start = next_start
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-amplitude/source_amplitude/api.py:172]: This change is not needed.

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/api.py [hunk 2]
```diff
@@ -203,6 +205,9 @@ def request_params(self, stream_slice: Mapping[str, Any], **kwargs) -> MutableMa
         params["end"] = pendulum.parse(stream_slice["end"]).strftime(self.date_template)
         return params
 
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-amplitude/source_amplitude/api.py:208]: Please leave only this change, this is sufficient for the problem.

**Review by bazarnov — APPROVED** [review]
LTGM with small correction, please check the review comments.

**Comment by b4stien** [comment]
@bazarnov I rewrote the change to `Events.stream_slices()` to be smaller. You're right it's not needed to fix the infinite loop, but it's still an issue regardless.

With the code as it is on master we end up with duplicate events at each slice boundaries. For instance with `[('20210101T00', '20210104T00'), ('20210104T00', '20210107T00'), ...]` the events of `20210104T00:00:00Z` to `20210104T01:00:00Z` will be fetched and stored twice, and as there is no mechanism to de-duplicate them afterwards they will also be in the destination twice.

Having duplicates in a "Full refresh - Overwrite" mode seems strange to me, although if you still want me to remove this change or extract it to another PR just tell me.

**Comment by bazarnov** [comment]
> @bazarnov I rewrote the change to `Events.stream_slices()` to be smaller. You're right it's not needed to fix the infinite loop, but it's still an issue regardless.
> 
> With the code as it is on master we end up with duplicate events at each slice boundaries. For instance with `[('20210101T00', '20210104T00'), ('20210104T00', '20210107T00'), ...]` the events of `20210104T00:00:00Z` to `20210104T01:00:00Z` will be fetched and stored twice, and as there is no mechanism to de-duplicate them afterwards they will also be in the destination twice.
> 
> Having duplicates in a "Full refresh - Overwrite" mode seems strange to me, although if you still want me to remove this change or extract it to another PR just tell me.

Got it, thanks for detailed explanation.

**Comment by bazarnov** [comment]
@marcosmarxm 
I think, we could proceed further with this contribution.

**Comment by harshithmullapudi** [comment]
Thanks @bazarnov

**Comment by b4stien** [comment]
Hi @harshithmullapudi thanks for the fixes!

**Pull request merged**